### ISSUE TITLE
Implement StackedGraphCast and StackedDataset

### ISSUE THREAD

**Issue by timothyas** [NOAA-PSL/graph-ufs#44]
We want to operate in the stacked / channel space rather than the expanded variable space. That is we want to be in (up to dimension order):
```
[sample, time, lat, lon, channels] # or [sample, time, x, channels]
```
instead of an expanded dataset where each variable is separated with dimsnions like
```
[sample, time, level, lat, lon]
```

We want to do this for several reasons:
1. Right now we're assuming we need to train with an "intermediate" zarr store that is ready to go for training. This should be in the channel space so that we're storing and working with chunks that are `{sample=1, time=-1, lat=-1, lon=-1, channels=-1}`, i.e. each data sample is a single chunk. This should improve I/O, and working in this channel space makes it easier to explore other data types like pytorch tensors or things like that.
2. This removes the need to do all the stacking and subsampling on the fly
3. If we do end up going with "pulling data on the fly" as in [this blogpost](https://earthmover.io/blog/cloud-native-dataloader) then we'll probably want to work with either the pytorch or tensorflow dataloaders in order to handle the prefetching (among other things). These data loaders need to work with more basic data types than xarray datasets, e.g. numpy arrays or pytorch tensors
4. Basically all other ML workflows expect this stacked channel space view of the data, so this is where we will want to be to explore other architectures more readily.

This involves changes to the GraphCast code and to our data loading.

## StackedGraphCast

Because of the way GraphCast is written, this actually involves quite a few steps:

- [x] StackedGraphCast (just the prediction)
- [x] Stacked loss function
- [x] Stacked bfloat16 casting
- [x] StackedInputsAndResiduals (normalization and dealing with residual increment during time stepping)
- [x] Deal with loss function weighting
     - [x] Check loss function types (both Typing and actual values) with the various additions below and general cleanup
- [x] Clean up the stacked predictor class inheritance
- [x] Make sure it works with batch_size > 1
- [x] Make the task_config stuff backward compatible within emulator class

### Future work: multi step autoregressive rollouts

Actually, we have really focused on multiple time steps, except maybe in the model evaluation. I remember making the assumption in the original data generator that we wanted a single time step for the target lead times, so I'm assuming that we will have a good deal of code edits for this. From the StackedGraphcast side this will entail:

1. Stacked autoregressive Predictor wrapper
2. Stacked rollout.py

However, note that we can use the same model parameters for stacked or regular mode. So we can train in "stacked mode" and just use the original graphcast implementation for multi step rollout predictions. Since we are not implementing multi step rollout training in p1, I'll table this for now.

## Stacked data loading

We could do this by reworking our current home grown data loading/generators, but I wanted to experiment with tools like xbatcher and the pytorch data loader. So far it's very clean. Items to do here:

- [x] Implement numpy-like stacked generator
- [x] Make it so we can also grab xarrays to easily use the original graphcast implementation (for comparison)
- [x] Integrate Dataset with Dataloader
